docs/overview.rst

```diff
@@ -1,11 +1,410 @@
+
+>>> from fuel.streams import DataStream
+>>> data_stream = DataStream(dataset=dataset, iteration_scheme=scheme)
+>>> for data in data_stream.get_epoch_iterator():
+...     print(data[0].shape, data[1].shape)
+(4, 2, 2) (4, 1)
+(4, 2, 2) (4, 1)
+
+
+Transformers: apply some transformation on the fly
+--------------------------------------------------
+
+.. admonition:: In summary
+    :class: tip
+
+    * :class:`Transformer` subclasses are responsible for taking data stream(s)
+      as input and producing a data stream as output, which applies some
+      transformation to the input stream(s).
+    * Transformers can be chained together to form complex data processing
+      pipelines.
+
+Some data streams take data streams as input. We call them *transformers*, and
+they enable us to build complex data preprocessing pipelines.
+
+Transformers are :class:`Transformer` subclasses. Most of the the transformers
+you'll encounter are located in the ``fuel.transformers`` module. Here are some
+commonly used ones:
+
+* :class:`Flatten`: flattens the input into a matrix (for batch input) or a
+  vector (for examplewise input).
+* :class:`ScaleAndShift`: scales and shifts the input by scalar quantities.
+* :class:`Cast`: casts the input into some data type.
+
+As an example, let's standardize the images we have by substracting their mean
+and dividing by their standard deviation.
+
+>>> from fuel.transformers import ScaleAndShift
+>>> # Note: ScaleAndShift applies (batch * scale) + shift, as
+>>> # opposed to (batch + shift) * scale.
+>>> scale = 1.0 / features.std()
+>>> shift = - scale * features.mean()
+>>> standardized_stream = ScaleAndShift(data_stream=data_stream,
+...                                     scale=scale, shift=shift,
+...                                     which_sources=('features',))
+
+The resulting data stream can be used to iterate over the dataset just like
+before, but this time features will be standardized on-the-fly.
+
+>>> for batch in standardized_stream.get_epoch_iterator():
+...     print(batch)
+(array([[[ 0.18530572, -1.54479571],
+        [ 0.42249705,  0.24111545]],
+<BLANKLINE>
+       [[-1.30760439,  0.98059429],
+        [-1.43317627, -1.2238898 ]],
+<BLANKLINE>
+       [[ 1.46892937,  1.58054882],
+        [ 0.47830677, -1.2657471 ]],
+<BLANKLINE>
+       [[ 0.63178351, -0.28907693],
+        [-0.40069638,  1.10616617]]]), array([[1],
+       [0],
+       [3],
+       [2]]))
+(array([[[ 1.32940506, -0.2332672 ],
+        [-1.60060544, -0.31698179]],
+<BLANKLINE>
+       [[ 0.03182898,  0.50621164],
+        [-1.64246273,  1.28754777]],
+<BLANKLINE>
+       [[ 0.88292727, -0.34488665],
+        [ 0.15740086,  1.51078666]],
+<BLANKLINE>
+       [[-1.00065091, -0.84717417],
+        [ 0.84106998, -0.19140991]]]), array([[2],
+       [0],
+       [3],
+       [2]]))
+
+Now, let's imagine that for some reason (e.g. running Theano code on GPU) we
+need features to have a data type of ``float32``.
```

Comment:
[...] need features to have a data type of ``float32``. We can cast them on-the-fly with a :class:`ScaleAndShift` `Transformer`